napari/components/style.py

```diff
@@ -0,0 +1,74 @@
+from typing import Dict
+
+from ..utils.events.dataclass import evented_dataclass
+from ..utils.theme import palettes
+
+DEFAULT_THEME = 'dark'
+CUSTOM_THEME = 'custom'
+
+
+@evented_dataclass
+class Style:
+    """Style object with style information for the viewer.
+
+    Parameters
+    ----------
+    pallete : dict
+        Color pallete for the viewer
+    theme : str
+        Theme for the viewer.
+
+    Attributes
+    ----------
+    available_themes : dict of str: dict of str: str
+        Available palettes indexed by theme.
```

Comment:
Can you remind me why we have both themes and palettes? They seem to be more or less identical concepts to me...

Reply:
Yeah I'd like to see that changed. I think palettes referred to the dict of available themes and their colors, and "theme" was the "current" theme (ie a specific key in the palettes).  Maybe they can both use a variant on "theme"?

Reply:
yes, i find this confusing too, will have a think on what the best approach might be ...

Reply:
how about just "theme" and "available_themes" (i.e. just change `palettes`) ... you've already added here:
```python
    @property
     def available_themes(self):
         """dict of str: dict of str: str. Preset color palettes, indexed by theme."""
         return palettes 
```
so might as well make it official?

Reply:
I also don't think we even need palettes to be accessible from this object, palettes are a global concept, not instance specific, and can be modified as described in https://forum.image.sc/t/custom-theme-dictionary/44808/4?u=talley

Reply:
yeah... actually, now that I look more at the rest of this PR, I can see that this is a similar thing to the discussion we had over in #2005 .
I maintain that the viewer should simply have a theme (I don't like `viewer.style.theme`) and we don't need palettes to be accessible from the viewer at all.  It doesn't bother me at all that the logic for actually changing a theme is 20 lines.  If you _really_ hate having the logic that changes the theme on the viewer, you could make it a function called during `_on_theme_set()`, but to me, this PR seems like indirection that is not yet warranted

Reply:
> how about just "theme" and "available_themes" (i.e. just change palettes) ... you've already added here:

just so that it's clear i understand (as it clearly is confusing now), you're suggesting we keep
 - `theme`, the string of the current theme
-  `palette`, the colors inside the current theme (right now a dict, but could be a named tuple, see comment in main thread)
- `available_themes`, the tuple of strings of available themes

We don't expose
- `palettes`, the dict of palettes indexed by theme which is a global concept?

Reply:
not quite.  I'm saying `palette` and `available_themes` are the exact same thing, and only one should remain, preferably `available_themes` (and palette gets killed).  And yes, we don't `available_themes` (aka `palette`) on the viewer object itself at all.  If someone wants to modify the themes that are available to napari, then the use something like `napari.utils.theme.palettes['my_new_theme']` and then set it.
I would also say we could just check to see whether viewer.theme is a dict in the forms of `{'theme_name': {...<new_theme_dict>}}` then we'll go ahead and add their new theme to the global options and activate it... basically how we do it for colormaps

Reply:
Ok, playing this back to see I get it.

> I'm saying palette and available_themes are the exact same thing, and only one should remain, preferably available_themes (and palette gets killed).

Sounds like given the below definition of `available_themes` (which is different from what I just said above, but is what is in the PR now)

```python
    @property
     def available_themes(self):
         """dict of str: dict of str: str. Preset color palettes, indexed by theme."""
         return palettes 
```

then

```python
palette = available_themes[theme]
```

And so you're saying have `palette` with `available_themes` is redundant, because you can get the one so easily from the other.

Then you're further saying `available_themes` is global and so shouldn't be on the object and so now we could just have `viewer.theme` which is a string, and `viewer.palette` wouldn't exist anymore either.

This all makes sense. If there wasn't `available_themes` on the object/ viewer anymore, we could make `viewer._palette` a private/ read only attribute so that objects that were using the viewer could still get read access to it rather than having to import `available_themes`, but if you wanted to add a new theme you'd have to use the approach you suggest.

How does that sound?

Reply:
yeah, I think we're on the same page.

> we could make viewer._palette a private/ read only attribute so that objects that were using the viewer could still get read access to it rather than having to import available_themes

This is fine if you want, but most of the cases where this is being done (like `get_key_bindings_summary`, and setting the canvas background) should probably not be taking explicit colors anyway.  Qt has great theme propagation from parent to child widget (which is why we don't have a lot of explicit theme setting throughout the code base), so the few places where we have resorted to manually setting a color to something from the current palette really should be refactored out so that everything cascades our style sheets (or at least _some_ single source, rather that setting styles explicitly inline).  In light of that, I'd personally say go ahead and import from `napari.utils.themes` where necessary, since eventually it shouldn't be happening at all.

Reply:
> so the few places where we have resorted to manually setting a color to something from the current palette really should be refactored out so that everything cascades our style sheets (or at least some single source, rather that setting styles explicitly inline). In light of that, I'd personally say go ahead and import from napari.utils.themes where necessary, since eventually it shouldn't be happening at all.

That's a very fair point.